keps/sig-network/2004-topology-aware-routing/README.md

```diff
@@ -0,0 +1,603 @@
+
+We propose a new Service annotation: `service.kubernetes.io/same-zone`
+with three options:
+- `Random`: This represents the existing approach that is already used, no
+  special mechanism will be added. Traffic will be evenly distributed among
+  endpoints behind a Service.
+- `Require`: EndpointSlice controller will conduct a simple but aggressive
+  routing mechanism. It will group endpoints in a zone to an EndpointSlice Group
+  that only receives traffic from that zone. `Require` requires users to ensure
+  that a sufficient number of endpoints for a Service exist in each zone.
+  Services using this approach will only be reachable from zones that contain
+  endpoints for the Service.
+- `Prefer`: EndpointSlice controller will conduct a more conservative routing
+  mechanism. It will keep traffic in zone as long as each zone has a sufficient
+  number of endpoints. If a zone does not have a sufficient number of endpoints,
+  traffic will be routed to other zones. This will be explained in more detail
+  below.
+
+This same pattern could be used for topology aware routing at node and region
+level. We could add new annotations `service.kubernetes.io/same-node` and
+`service.kubernetes.io/same-region` for node and region respectively with
+the same values `Require` and `Prefer`. With all these three levels combined
+together, their individual value could be influenced by values of others, i.e.
+`Require` node would also `Require` zone and region.
+
+#### Controller Configuration
+
+In addition to the Service-level annotation, a global default can be configured
+on the EndpointSlice controller. This will define the behavior of the controller
+for Services without an annotation specified. For backwards compatibility this
+will initially default to the `Random` approach.
+
+### Notes/Constraints/Caveats
+
+In the future, if there is sufficient need, we may introduce another parameter
+for the `Prefer` approach that would allow users to specify a starting
+threshold. Starting threshold defines the starting point for `Prefer` (more
+detailed description below). We currently set the starting threshold to 3x the
+number of zones based on results from our test dataset for safer distribution
+results.
+
+### Risks and Mitigations
+
+- Extra complexity to EndpointSlice Controller introduced by `Require` and
+  `Prefer` could result in slower EndpointSlices update and potentially end up
+  limiting scalability.
+- When users opt-in to the `Require` routing scheme, it is their responsibility
+  to make sure the workloads are balanced.
+- In a scenario where all traffic originates from a single zone, with `Require`
+  and `Prefer` options, there is a chance that endpoints in that zone will be
+  overloaded while endpoints in other zones receive little to no traffic.
+  Without some sort of feedback (out of scope) this will not self-rectify.
+
```

Comment:
@robscott I have this doubt, can a pod in ZoneB access a Service with `service.kubernetes.io/same-zone: Require` that only has endpoints in ZoneA?
I mean,  the traffic load "service-endpoint" will be withing the zone, but at least in kube-proxy the ClusterIP -> Endpoint translation is done in the same node used by the pod client. how this traffic will be redirected to the other zone?

Reply:
It would not be able to. That approach relies on the user distributing workloads across all zones they intend to send traffic to. This could work well with some cross-sig work that would make this kind of scheduling easier and potentially update autoscaling to also take topology into account. I don't imagine this being a commonly used option at the beginning but I think it's still valuable for power users. I'll work on a better explanation for this in the KEP.

Reply:
I've added some better clarification around this to the KEP now.